content/200-concepts/100-components/02-prisma-client/260-client-extensions/100-model.mdx

```diff
@@ -133,3 +133,110 @@ prisma.$extends({
 ```
 
 You cannot use `Prisma.getExtensionContext(this)` inside an [`$allModels`](#add-a-custom-method-to-all-models-in-your-schema) structure.
+
+## Advanced type safety on models
+
+<Admonition type = "info">
+
+This feature is available from version 4.9.0.
+
+</Admonition>
+
+**Rewrite this intro on Monday:** To help with type safety, Prisma provides code validation for the internal types in generic models. It is especially useful for published extensions, because other users can more easily benefit from autocompletion, no matter what models apply. **TBA:** link to "sharing extensions" section; also link here from that section, but make it clear in the link text that this is an advanced topic
+
+The following type utilities are available:
+
+- `Exact`: Assists with strict type safety - narrows the input down. **TBA** - clarify what this specifically does - does it declare the objects/arguments that you want to apply this type safety to?
+- `Args`: Dynamically retrieves the input type of a given model and a given operation. When you change the model, autocomplete indicates whether the fields ezists on it.
+- `Result`: Takes the input types and provides the result for a given model and a given operation.
+- `Payload`: Retrieves the entire structure of the result, as scalars and relations objects.
+
+- **TBA:** some more code comments to annotate the following example, especially in the $query section
+
+```ts
+const prisma = new PrismaClient()
+
+const xprisma = prisma.$extends({
+  model: {
+    $allModels: {
+      findOrCreate<T, A>(
+        // On the following line, `this` and `T` refer to a generic model,
+        // which is a placeholder for the model that is in use.
+        // For example, `this` might return `prisma.user`.
```

Comment:
Should this be in proper text somewhere or is it fine here in the example?

Reply:
I've reworked the code comments now, so I'll resolve this convo